README.md

```diff
@@ -21,30 +24,29 @@ to help analysis of the program execution and performance.
 Features
 ========
 
-It traces each function in the executable and shows time duration.  It
-can also trace external library calls - but usually entry and exit are
-supported.  Optionally it's possible to trace other (nested) external
+The tool traces each function in the executable and shows time durations.  It
+can also trace external library calls - but usually only entry and exit are
```

Comment:
I think this part can be improved further.  I meant it only records PLT calls in the main executable so that we can see total library call durations.

Reply:
Thanks, I've pushed a new update of this paragraph with this new content:

uftrace traces each function in the executable and shows time durations.
By hooking into the procedure linkage tables of shared libraries, it can
also trace shared library calls like the `ltrace` tool, but with a nicer,
more readable and colored output format which even includes resolving of
C FILE pointers for stdin, stdout and stderr.

Reply:
Hi @bernhardkaindl, thanks very much for your help!

I couldn't fully follow the changes, but this part doesn't include explanation of `--nest-libcall`, which allows uftrace to follow nested library function calls. `ltrace` doesn't support this as far as I know.

Reply:
You can see the output difference of `ltrace` and `uftrace`.
```
$ ltrace -C clang 2>&1 | head
getenv("bar")                                    = nil
clang::noteBottomOfStack()(1, 0x7ffd3719af78, 0x7ffd3719af88, 0x420c78) = 0x7fd75a6d8778
llvm::InitLLVM::InitLLVM(int&, char const**&, bool)(0x7ffd3719ad00, 0x7ffd37199a64, 0x7ffd37199a58, 1) = 0
llvm::setBugReportMsg(char const*)(0x41df93, 0, 0, 1) = 0
memcpy(0x7ffd3719a500, "^\303\0317\375\177\0\0", 8) = 0x7ffd3719a500
llvm::sys::Process::FixupStandardFileDescriptors()(0x7ffd3719a500, 0x7ffd3719c35e, 8, 0x7ffd3719c35e) = 0
LLVMInitializeAArch64TargetInfo(2, 0x7fd7537d846f, 0x7fd753c26220, 0x7fd753713eee) = 0x7fd75a6a5908
LLVMInitializeAMDGPUTargetInfo(0x7fd75a6a5a08, 0x7fd757a834ee, 0x7fd757b82979, 0x7fd757b006eb) = 0x7fd75a6aba18
LLVMInitializeARMTargetInfo(0x7fd75a6abb18, 0x7fd757a87ab0, 0x7fd757b9ee85, 0x7fd757b86b9f) = 0x7fd75a6af5e0
LLVMInitializeAVRTargetInfo(0x7fd75a6af6e0, 0x7fd757a834ce, 0x7fd757bcf606, 0x7fd757b4cf5e) = 0x7fd75a6af6e0

$ uftrace --nest-libcall -a clang | head -12
clang: error: no input files
# DURATION     TID     FUNCTION
 292.220 us [ 82779] | getenv("bar") = "NULL";
            [ 82779] | clang::noteBottomOfStack() {
   1.160 us [ 82779] |   __tls_get_addr();
   7.834 us [ 82779] | } /* clang::noteBottomOfStack */
            [ 82779] | llvm::InitLLVM::InitLLVM() {
   0.570 us [ 82779] |   pthread_mutex_lock(0x7f04dfc50090) = 0;
            [ 82779] |   operator new(48) {
   0.297 us [ 82779] |     malloc(48) = 0x19f9720;
   1.596 us [ 82779] |   } = 0x19f9720; /* operator new */
   0.184 us [ 82779] |   pthread_mutex_unlock(0x7f04dfc50090) = 0;
   0.152 us [ 82779] |   pthread_mutex_lock(0x19f9720) = 0;
```
Especially, uftrace shows the library calls in more depth

Reply:
Thanks, I didn't even know about this nice combination of --nest-libcall and -a!
I guess a dedicated EXAMPLES.md which shows off each feature would be nice!

_**Do you accept this correction (below) of my omission? Pushed now.**_

Example with nested library calls:
```py
uftrace -la kdemv -h
# DURATION     TID     FUNCTION
   0.598 us [2064374] | std::ios_base::Init::Init();
   0.102 us [2064374] | __cxa_atexit();
            [2064374] | QApplication::QApplication() {
            [2064374] |   operator new(360) {
   0.263 us [2064374] |     malloc(360) = 0x55eddb5fad70;
   2.620 us [2064374] |   } = 0x55eddb5fad70; /* operator new */
            [2064374] |   QApplicationPrivate::QApplicationPrivate() {
            [2064374] |     QGuiApplicationPrivate::QGuiApplicationPrivate() {
            [2064374] |       QCoreApplicationPrivate::QCoreApplicationPrivate() {
   2.178 us [2064374] |         QObjectPrivate::QObjectPrivate();
   0.250 us [2064374] |         QReadWriteLock::QReadWriteLock();
   0.328 us [2064374] |         geteuid() = 1000;
   0.218 us [2064374] |         getuid() = 1000;
```

Direct link to the updated README.md:
https://github.com/bernhardkaindl/uftrace/blob/improve-README.md/README.md

Reply:
Thanks for the update. It looks good but I guess it might be useful if you can add some examples with string argument printed together as it looks nicer.

Reply:
For example, this kind of output looks good as follows.
```py
$ uftrace record -la lsusb > /dev/null 
$ uftrace replay | head -30
# DURATION     TID     FUNCTION
 299.951 us [ 23303] | setlocale(LC_TYPE, "") = "en_US.UTF-8";
   1.434 us [ 23303] | getopt_long(1, 0x7ffcd0a4ea68, "D:vtP:p:s:d:Vh") = -1;
            [ 23303] | udev_new() {
   0.301 us [ 23303] |   malloc(16) = 0x55aef04df990;
   2.143 us [ 23303] | } /* udev_new */
            [ 23303] | udev_hwdb_new() {
   0.424 us [ 23303] |   calloc(1, 200) = 0x55aef04df9b0;
   5.724 us [ 23303] |   fopen64("/etc/systemd/hwdb/hwdb.bin", "re") = 0;
   0.069 us [ 23303] |   __errno_location() = 0x7fb68c6d56f8;
   0.231 us [ 23303] |   strlen("/etc/systemd/hwdb/hwdb.bin") = 26;
   2.068 us [ 23303] |   fopen64("/etc/udev/hwdb.bin", "re") = 0;
   0.056 us [ 23303] |   __errno_location() = 0x7fb68c6d56f8;
   0.111 us [ 23303] |   strlen("/etc/udev/hwdb.bin") = 18;
   1.989 us [ 23303] |   fopen64("/usr/lib/systemd/hwdb/hwdb.bin", "re") = 0;
   0.047 us [ 23303] |   __errno_location() = 0x7fb68c6d56f8;
   0.117 us [ 23303] |   strlen("/usr/lib/systemd/hwdb/hwdb.bin") = 30;
   1.940 us [ 23303] |   fopen64("/lib/systemd/hwdb/hwdb.bin", "re") = 0;
   0.049 us [ 23303] |   __errno_location() = 0x7fb68c6d56f8;
   0.116 us [ 23303] |   strlen("/lib/systemd/hwdb/hwdb.bin") = 26;
   2.596 us [ 23303] |   fopen64("/lib/udev/hwdb.bin", "re") = 0x55aef02ab740;
   0.185 us [ 23303] |   fileno(0x55aef02ab740) = 4;
   1.375 us [ 23303] |   fstat64();
   0.080 us [ 23303] |   fileno(0x55aef02ab740) = 4;
   3.973 us [ 23303] |   mmap64(0, 0xb0c1e7, PROT_READ, MAP_SHARED, 4, 0) = 0x7fb689800000;
   0.142 us [ 23303] |   malloc(24) = 0x55aef04dfa80;
   0.084 us [ 23303] |   malloc(24) = 0x55aef04dfaa0;
  31.215 us [ 23303] | } /* udev_hwdb_new */
            [ 23303] | libusb_init() {
   1.118 us [ 23303] |   pthread_mutex_lock(0x7fb68cb7a420) = 0;
```
It's off topic but we can improve the output  by handling `__errno_location`.

The prototype of `__errno_location` might be this.
```
int *__errno_location(void);
```
But uftrace needs to add support for dereferencing a type, then it can be shown as follows.
```
__errno_location() = 2;
```
Then we can give hint for the `errno` values from `errno -l`. Maybe I can create an issue for this.

Reply:
Thanks, I've added it like this:
```py
$ uftrace record -la lsusb >/dev/null 
$ uftrace replay -f+module
or simply:
$ uftrace -la -f+module lsusb  # -f+module adds the module name column(man uftrace-live):
# DURATION     TID        MODULE NAME   FUNCTION
 306.339 us [ 23561]            lsusb | setlocale(LC_TYPE, "") = "en_US.UTF-8";
   1.163 us [ 23561]            lsusb | getopt_long(1, 0x7fff7175f6a8, "D:vtP:p:s:d:Vh") = -1;
            [ 23561]            lsusb | udev_new() {
   0.406 us [ 23561] libudev.so.1.7.2 |   malloc(16) = 0x55e07277a7b0;
   2.620 us [ 23561]            lsusb | } /* udev_new */
            [ 23561]            lsusb | udev_hwdb_new() {
   0.427 us [ 23561] libudev.so.1.7.2 |   calloc(1, 200) = 0x55e07277a7d0;
   5.829 us [ 23561] libudev.so.1.7.2 |   fopen64("/etc/systemd/hwdb/hwdb.bin", "re") = 0;
```